src/urllib3/contrib/securetransport.py

```diff
@@ -570,40 +570,49 @@ def recv_into(
         if nbytes is None:
             nbytes = len(buffer)
 
-        buffer = (ctypes.c_char * nbytes).from_buffer(buffer)
         processed_bytes = ctypes.c_size_t(0)
+        processed_bytes_total = 0
 
-        with self._raise_on_error():
-            result = Security.SSLRead(
-                self.context, buffer, nbytes, ctypes.byref(processed_bytes)
-            )
+        while nbytes:
+            buffer = (ctypes.c_char * nbytes).from_buffer(buffer, processed_bytes.value)
 
-        # There are some result codes that we want to treat as "not always
-        # errors". Specifically, those are errSSLWouldBlock,
-        # errSSLClosedGraceful, and errSSLClosedNoNotify.
-        if result == SecurityConst.errSSLWouldBlock:
-            # If we didn't process any bytes, then this was just a time out.
-            # However, we can get errSSLWouldBlock in situations when we *did*
-            # read some data, and in those cases we should just read "short"
-            # and return.
-            if processed_bytes.value == 0:
-                # Timed out, no data read.
-                raise socket.timeout("recv timed out")
-        elif result in (
-            SecurityConst.errSSLClosedGraceful,
-            SecurityConst.errSSLClosedNoNotify,
-        ):
-            # The remote peer has closed this connection. We should do so as
-            # well. Note that we don't actually return here because in
-            # principle this could actually be fired along with return data.
-            # It's unlikely though.
-            self.close()
-        else:
-            _assert_no_error(result)
+            with self._raise_on_error():
+                result = Security.SSLRead(
+                    self.context, buffer, nbytes, ctypes.byref(processed_bytes)
+                )
+
+            nbytes -= processed_bytes.value
+            processed_bytes_total += processed_bytes.value
+
+            # There are some result codes that we want to treat as "not always
+            # errors". Specifically, those are errSSLWouldBlock,
+            # errSSLClosedGraceful, and errSSLClosedNoNotify.
+            if result == SecurityConst.errSSLWouldBlock:
+                # If we didn't process any bytes, then this was just a time out.
+                # However, we can get errSSLWouldBlock in situations when we *did*
+                # read some data, and in those cases we should just read "short"
+                # and return.
+                if processed_bytes.value == 0:
```

Comment:
Should this be `if processed_bytes_total == 0`? If we issue two reads, one that returns data and one that doesn't return data w/ errSSLWouldBlock then we want to return the already read data instead of erroring with a timeout.

Reply:
I am not sure about this.

CPython 3.7 and 3.9 read data in a loop as I wrote in the description. If the second or subsequent iteration of the loop calls `urllib3.contrib.securetransport.WrappedSocket.recv_into`, gets to the check and finds it true, it will raise the timeout error.

My code moves the loop inside of `recv_into`. So I assume that erroring with a timeout  `if processed_bytes.value == 0:` is consistent with what CPython 3.7 and 3.9 do.

---

I am not sure if returning read data knowing that it is not complete is the right way because the user may not even know that the response is incomplete.

Reply:
Yes, there is no loop in the method, but `BufferedReader` is used when `SecureTransport` is not used: https://github.com/urllib3/urllib3/pull/2674#issuecomment-1183657472.

> You can see that the `socket.recv` implementation restricts the amount of data you can ask for in one read on Windows to 16K.

Isn't that 2**32 - 1 bytes?